### PR TITLE
Improve test coverage for CCS search cancellation and fix response bugs

### PR DESCRIPTION
**Pull request by quux00**
Fixes two bugs in the Async Search Response after searches have been cancelled.

Adds additional IT tests for cross-cluster async search cancellation - both by explicit cancellation 
via the Task API and by deleting the async search while it is still running.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java [hunk 1]
```diff
@@ -120,16 +120,11 @@ private void getSearchResponseFromTask(
...
-            if (task == null) {
+            if (task == null || task.isCancelled()) {
                 getSearchResponseFromIndex(searchId, request, nowInMillis, listener);
                 return;
             }
 
-            if (task.isCancelled()) {
-                listener.onFailure(new ResourceNotFoundException(searchId.getEncoded()));
-                return;
-            }
```
Comment (quux00) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java:131]: Based on my analysis and testing (both manual and automated), this fail-fast if block is not needed and actually undesirable.

When a long-running search is cancelled via the `_tasks/_cancel` API but before the task and search fully complete, the GET `_async_search/:id` endpoint throws this "resource_not_found_exception". But then when the task is unregistered and the search is fully finished, the GET `_async_search/:id` endpoint no longer throws that exception, but instead returns the expected output showing that correct error (that the search was cancelled).

With this change, that no longer happens. The GET `_async_search/:id` endpoint returns useful information (like the `GET `_async_search/status/:id` endpoint` already does) after the cancellation but before the search has fully shutdown, as shown below.

<details><summary><code>GET _async_search/:id</code> after task is cancelled but it is not yet unregistered and the search has not fully shutdown</summary>

```
{
  "id": "FkswMWdMc2VEUVJxTXNMVGtnaUliU3caMl90V1VTZHBUd1d4WEx5dDhwMlhCdzoxMjg=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1686667932790,
  "expiration_time_in_millis": 1687099932790,
  "response": {
    "took": 131,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 0,
    "_shards": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 0,
        "relation": "gte"
      },
      "max_score": "",
      "hits": []
    }
  }
}
```

</details>

<details><summary><code>GET _async_search/:id</code> after task is unregistered and the search has fully shutdown</summary>

```js
{
  "id": "FkswMWdMc2VEUVJxTXNMVGtnaUliU3caMl90V1VTZHBUd1d4WEx5dDhwMlhCdzoxMjg=",
  "is_partial": true,
  "is_running": false,
  "start_time_in_millis": 1686667932790,
  "expiration_time_in_millis": 1687099932790,
  "response": {
    "took": 36810,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 0,
    "_shards": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "failed": 3,
      "failures": [
        {
          "shard": 0,
          "index": "blogs",
          "node": "2_tWUSdpTwWxXLyt8p2XBw",
          "reason": {
            "type": "task_cancelled_exception",
            "reason": "task cancelled [by user request]"
          }
        }
      ]
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 0,
        "relation": "gte"
      },
      "max_score": "",
      "hits": []
    }
  },
  "error": {
    "type": "status_exception",
    "reason": "error while executing search",
    "caused_by": {
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
        {
          "shard": 0,
          "index": "remote1:blogs",
          "node": "tsYK9MXVRcezptK7HYwLLA",
          "reason": {
            "type": "task_cancelled_exception",
            "reason": "task cancelled [by user request]"
          }
        }
      ],
      "caused_by": {
        "type": "task_cancelled_exception",
        "reason": "task cancelled [by user request]"
      }
    }
  }
}
```

</details>

This behavior now matches how the `GET _async_search/status/:id` endpoint behaves under these conditions as well (not shown).

Comment (quux00) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java:131]: The other scenario to consider for this code change is when the search is deleted via `DELETE _async_search`. In that throwing `ResourceNotFoundException` is correct. However, in this case, there is also no need to have this short-circuit block, as the call below it to [sendFinalResponse](https://github.com/elastic/elasticsearch/blob/b2504a9f87cb854ce35214b23bda41d600cca2b2/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java#L133) will also throw `ResourceNotFoundException`.

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterSearchIT.java [hunk 2]
```diff
@@ -261,6 +297,15 @@ public void testCancel() throws Exception {
                 assertThat(transportService.getTaskManager().getBannedTaskIds(), Matchers.empty());
             }
         });
+
+        try {
+            queryFuture.result();
```
Comment (javanna) [inline server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterSearchIT.java:302]: use expectThrows?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java (hunk 1 above)
Comment (javanna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncResultsService.java:131]: thanks for sharing all the context!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java [hunk 3]
```diff
@@ -102,7 +102,7 @@ public void writeTo(StreamOutput out) throws IOException {
     }
 
     public AsyncSearchResponse clone(String searchId) {
-        return new AsyncSearchResponse(searchId, searchResponse, error, isPartial, false, startTimeMillis, expirationTimeMillis);
+        return new AsyncSearchResponse(searchId, searchResponse, error, isPartial, isRunning, startTimeMillis, expirationTimeMillis);
```
[the remark this answers is not in the document]
Comment (javanna) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java:105]: ++

File: x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+                    .get()
+                    .getTasks()
+                    .stream()
+                    .filter(t -> t.action().contains(SearchAction.NAME))
+                    .collect(Collectors.toList());
+                assertThat(remoteSearchTasks.size(), greaterThan(0));
+                remoteClusterSearchTasks.set(remoteSearchTasks);
+            });
+
+            for (TaskInfo taskInfo : remoteClusterSearchTasks.get()) {
```
Comment (javanna) [inline x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java:167]: should we check that the list is not empty, or better that there is one item per cluster?

File: x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
```
Comment (javanna) [inline x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java:293]: check that a certain number of tasks is expected?

**Review by javanna — APPROVED** [review]
left a couple of comments, LGTM otherwise. 

I think that we will have to expand cancellation tests to the case where we don't minimize roundtrips. There is this TODO (https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L353) that needs addressing in that area and probably not enough coverage. For the minimize roundtrips scenario though, this PR is great!

File: server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterSearchIT.java (hunk 2 above)
Comment (quux00) [reply server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterSearchIT.java:302]: Fixed.

File: x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java (hunk 4 above)
Comment (quux00) [reply x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java:167]: I think that's handled in the assertBusy: https://github.com/elastic/elasticsearch/pull/97029/files#diff-54ddb375db8a4150e4c6c453be159da93d883315d373d31c2a14d0a2baaccac9R163

It waits until the size is not zero.

In terms of checking for a size of 1, on the remote cluster there is often more than 1:

```
{
  "node" : "DA3ivEjkS76A4emXH3o2Wg",
  "id" : 80,
  "type" : "transport",
  "action" : "indices:data/read/search[phase/query]",
  "start_time" : "2023-06-30T13:51:16.978Z",
  "start_time_in_millis" : 1688133076978,
  "running_time" : "79.8ms",
  "running_time_in_nanos" : 79874042,
  "cancellable" : true,
  "cancelled" : false,
  "parent_task_id" : "msazZKLYS0apvopKa7On0A:66",
  "headers" : { }
}
{
  "node" : "msazZKLYS0apvopKa7On0A",
  "id" : 66,
  "type" : "transport",
  "action" : "internal:transport/proxy/indices:data/read/search[phase/query]",
  "start_time" : "2023-06-30T13:51:16.977Z",
  "start_time_in_millis" : 1688133076977,
  "running_time" : "80.9ms",
  "running_time_in_nanos" : 80931416,
  "cancellable" : true,
  "cancelled" : false,
  "parent_task_id" : "K9DpCDRmRl2ajqALS4DNig:83",
  "headers" : { }
}
```

So I put it into a loop to assert that however many there are, they are not yet cancelled, which is the main concern for the test.

File: x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java (hunk 5 above)
Comment (quux00) [reply x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java:293]: See comment above.  The assertBusy ensures there is at least 1. There is often more than one with a `internal:transport/proxy/indices:data/read/search[phase/query]` which seems like a transport implementation detail that I didn't want to make asserts about.

**Pull request merged**